scaler/client/agent/future_manager.py

```diff
@@ -55,37 +58,57 @@ def on_task_result(self, result: TaskResult):
 
             profile_result = retrieve_profiling_result_from_task_result(result)
 
-            try:
-                if result.status == TaskStatus.NotFound:
-                    future.set_exception(TaskNotFoundError(f"task not found: {task_id.hex()}"), profile_result)
-                    return
-
-                if result.status == TaskStatus.WorkerDied:
+            match result.result_type:
+                case TaskResultType.FailedWorkerDied:
                     future.set_exception(
                         WorkerDiedError(f"worker died when processing task: {task_id.hex()}"), profile_result
                     )
-                    return
-
-                if result.status == TaskStatus.NoWorker:
-                    future.set_exception(
-                        NoWorkerError(f"no available worker when processing task: {task_id.hex()}"), profile_result
-                    )
-                    return
 
-                if result.status == TaskStatus.Canceled:
-                    future.set_exception(DisconnectedError("client disconnected"), profile_result)
-                    return
+                case TaskResultType.Success:
+                    result_object_id = self.__get_result_object_id(result)
+                    future.set_result_ready(result_object_id, TaskState.Success, profile_result)
 
-                if result.status in {TaskStatus.Success, TaskStatus.Failed}:
-                    assert len(result.results) == 1
-                    result_object_id = ObjectID(result.results[0])
-                    future.set_result_ready(result_object_id, result.status, profile_result)
-                    return
+                case TaskResultType.Failed:
+                    result_object_id = self.__get_result_object_id(result)
+                    future.set_result_ready(result_object_id, TaskState.Failed, profile_result)
 
-                raise TypeError(f"Unknown task status: {result.status}")
-            except InvalidStateError:
-                return  # Future got canceled
+                case _:
+                    raise TypeError(f"{result.task_id.hex()}: Unknown task status: {result.result_type}")
 
-    def on_cancel_task(self, task_cancel: TaskCancel):
+    def on_task_cancel_confirm(self, cancel_confirm: TaskCancelConfirm):
         with self._lock:
-            self._task_id_to_future.pop(task_cancel.task_id, None)
+            task_id = cancel_confirm.task_id
+            if task_id not in self._task_id_to_future:
+                return
+
+            future = self._task_id_to_future.pop(task_id)
+            assert cancel_confirm.task_id == future.task_id
+
+            match cancel_confirm.cancel_confirm_type:
+                case TaskCancelConfirmType.Canceled:
+                    future.set_canceled()
+
+                case TaskCancelConfirmType.CancelNotFound:
+                    logging.error(f"{task_id!r}: task not found")
+                    future.set_canceled()
```

Comment:
Could be better to make the future raise an exception instead.

```suggestion
                    future.set_exception(InvalidStateError())
```